packages/animations/lib/src/composite_animation_widget.dart

```diff
@@ -0,0 +1,270 @@
+// Copyright 2019 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+
+/// Function signature for building component transitions (forward and reverse)
+/// for [CompositeAnimationWidget].
+///
+/// The function should return a widget which wraps the given `child`.
+typedef ComponentTransitionBuilder = Widget Function(Widget, Animation<double>);
+
+/// A widget that animate it's child with different transitions based on
+/// [AnimationStatus] of [CompositeAnimationWidget.animation].
+/// If the ([CompositeAnimationWidget.animation]) value goes forward,
+/// only [CompositeAnimationWidget.forwardTransitionBuilder] animates
+/// from 0.0 to 1.0 and If goes reverse, only
+/// [CompositeAnimationWidget.reverseTransitionBuilder] animates from 1.0 to 0.0.
+///
+/// The following example shows how use this widget to compose different
+/// animations when going forward or reverse driven by one [Animation] object
+/// only. Fading in on forward and Scaling down on reverse.
+///
+/// ```dart
+/// CompositeAnimationWidget(
+///   animation: animation,
+///   forwardTransitionBuilder: (Widget child, Animation<double> animation)
+///     => FadeTransition(
+///       opacity: animation,
+///       child: child,
+///     ),
+///   reverseTransitionBuilder: (Widget child, Animation<double> animation)
+///     => ScaleTransition(
+///       scale: animation,
+///       child: child,
+///     ),
+///   child: child,
+/// )
+/// ```
+///
+/// It is used in [FadeThroughTransition], [SharedAxisTransition],
+/// and [FadeScaleTransition] widgets.
+///
+/// This widget also preserve it's child state by building it's transitions by
+/// composition and only changing animation values of both transitions instead
+/// of changing the shape of widget subtree levels in [AnimatedBuilder]
+/// based on [AnimationStatus] that causes the state object in that particular
+/// subtree level to get recreated.
+///
+/// To ensure the state of the child widget will survive, don't define
+/// conditional logic to switch subtrees in both transition builders required
+/// by this widget.
+class CompositeAnimationWidget extends StatefulWidget {
+  /// Creates a widget that animate it's child with different transitions based
+  /// on [AnimationStatus] of [animation].
+  ///
+  /// If the ([animation]) value goes forward, only [forwardTransitionBuilder]
+  /// animates from 0.0 to 1.0 and If goes reverse, only [reverseTransitionBuilder]
```

Comment:
I think it will be easier to reason about the animations if they go from 0.0 to 1.0 for both builders. Let's say you want to change the background from green to white on exit. If animations go from 0.0 to 1.0 all you have to do is define a `Tween(begin: Colors.green, end: Colors.white)`. If they go from 1.0 to 0.0 you have to switch begin/end around, which is awkward to think about. (This will require some changes to the existing animations, e.g. begin/end for some tweens need to be switched around and some curves may need to be flipped).